pilot/helpers/agents/CodeMonkey.py

```diff
@@ -1,38 +1,353 @@
+
+        if replace_complete_file:
+            content = self.replace_complete_file(
+                convo,
+                standalone,
+                code_changes_description,
+                file_content,
+                file_name, files
+            )
+            blocks = self.get_code_blocks(llm_response)
+
+        if content and content != file_content:
+            self.project.save_file({
+                'path': step['path'],
+                'name': step['name'],
+                'content': content,
+            })
+
+        convo.remove_last_x_messages(exchanged_messages)
         return convo
+
+    def replace_complete_file(
+        self,
+        convo: AgentConvo,
+        standalone: bool,
+        code_changes_description: str,
+        file_content: str,
+        file_name: str,
+        files: list[dict]
+    ) -> str:
+        """
+        As a fallback, replace the complete file content.
+
+        This should only be used if we've failed to replace individual code blocks.
+
+        :param convo: AgentConvo instance
+        :param standalone: True if this is a standalone conversation
+        :param code_changes_description: description of the code changes
+        :param file_content: content of the file being updated
+        :param file_name: name of the file being updated
+        :param files: list of files to send to the LLM
+        :return: updated file content
+
+        Note: if even this fails for any reason, the original content is returned instead.
+        """
+        llm_response = convo.send_message('development/implement_changes.prompt', {
+            "full_output": True,
+            "standalone": standalone,
+            "code_changes_description": code_changes_description,
+            "file_content": file_content,
+            "file_name": file_name,
+            "files": files,
+        })
+        blocks = self.get_code_blocks(llm_response)
+        if len(blocks) != 1:
+            logger.warning(f"Unable to implement code changes after 5 retries and a fallback: {code_changes_description}")
+            return file_content
+
+        return blocks[0]
+
+
+    def identify_files_to_change(self, code_changes_description: str, files: list[dict]) -> list[str]:
+        """
+        Identify files to change based on the code changes description
+
+        We really just want one file, but this can handle mutliple files
+        and will return them all, or an empty list if no files are
+        returned.
+
+        :param code_changes_description: description of the code changes
+        :param files: list of files to send to the LLM
+        :return: list of files to change
+        """
+        convo = AgentConvo(self)
+        llm_response = convo.send_message('development/identify_files_to_change.prompt', {
+            "code_changes_description": code_changes_description,
+            "files": files,
+        })
+        blocks = self.get_code_blocks(llm_response)
```

Comment:
is it a problem here if `blocks` is odd length, meaning response was not as expected and is not processed properly?

Reply:
GitHub doesn't let me know what this was referring to, but I believe this was for determining which file we need to change, which is now using JSON in 8e301fa1cee720bd789e977e630c8adf55676bb1